### ISSUE TITLE
Set up MFA with TOTP as preferred and SMS and secondary fallback method

### ISSUE THREAD

**Issue by tylerjames** [aws-amplify/aws-sdk-ios#4632]
**State your question**
I am able to set up MFA using either TOTP or SMS. What I would like is to encourage my users to set up TOTP as the primary method and to only rely on SMS as a fallback method in the event that they lose access to their authenticator app.

I have heard of a mysterious "SELECT_MFA_TYPE" flow but I cannot seem to achieve it. I created an account and added both TOTP and SMS based MFA to it but set neither as preferred. I understood that in this scenario I would be asked what MFA method to use as part of the sign-in flow.  
Instead I am simply receiving an error as sign in that looks like this:
```
▿ AWSMobileClientError
  ▿ unknown : 1 element
    - message : "nil: Error information not available."
```
Which is quite a useless error. 

Is there any way to set up TOTP and SMS at the same time?

**Which AWS Services are you utilizing?**
AWSCognito

**Provide code snippets (if applicable)**
I'm setting the MFA method using this function. If I set either of these methods to `preferredMfa = true` then only that method will be used. If I set both to false then I get this unknown error.
```swift
  func setUserMFAPreference(preferredMFA: PreferredMFA, token: String) async throws -> Bool {
    guard let request = AWSCognitoIdentityProviderSetUserMFAPreferenceRequest()
    else {
      throw .unknownError
    }
    
    if preferredMFA == .sms,
       let setting = AWSCognitoIdentityProviderSMSMfaSettingsType()
    {
      setting.enabled = true
      setting.preferredMfa = false
      request.smsMfaSettings = setting
    }
    else if preferredMFA == .authenticator,
            let setting = AWSCognitoIdentityProviderSoftwareTokenMfaSettingsType()
    {
      setting.enabled = true
      setting.preferredMfa = false
      request.softwareTokenMfaSettings = setting
    } else {
      throw .unknownError
    }
    
    request.accessToken = token
    
    return try await withCheckedThrowingContinuation { continuation in
      AWSCognitoIdentityProvider
        .default()
        .setUserMFAPreference(request)
        .continueWith { task in
          if let error = task.error {
            continuation.resume(throwing: error)
            return
          }
          if let result = task.result {
            print("Result: \(result)")
            continuation.resume(returning: true)
          }
          return nil
        }
    }
  }
```

**Environment(please complete the following information):**
 - SDK Version: 2.30.4
 - Dependency Manager: SPM
 - Swift Version : 5

**Device Information (please complete the following information):**
 - Device: iPhone 14 Pro simulator
 - iOS Version: 16.4
 - Specific to simulators: unknown

**Comment by harsh62** [comment]
@tylerjames So based on your description and error.. I see conflicting implementations.. So would you be able to share your implementation for signing in and how you are handling the error?
For setting user MFA preference, I see that you are using `AWSCognitoIdentityProvider` whereas the error shared is of type `AWSMobileClient`. 

`AWSMobileClient` does not support TOTP.

**Comment by tylerjames** [comment]
It actually kind of does support TOTP, or it at least knows that you have some MFA enabled.

Sign in function is like this:
```swift
func signIn(email: String, password: String) async throws -> LoginResult {
  try await withCheckedThrowingContinuation { [unowned mobileClient] continuation in
    mobileClient.signIn(username: email, password: password) { result, error in
      if let error {
        continuation.resume(throwing: error)
        return
      }
      
      guard let result else {
        continuation.resume(throwing: .emptyResponse)
        return
      }
      
      continuation.resume(returning: result)
    }
  }
}
```

If you enable MFA using TOTP like in my first post then switching on `result.signInState` will yield `.smsMfa`. We then show a screen where a user can enter the code generated by their authenticator app and we submit it with:
```swift
func send2FACode(code: String) async throws -> LoginResult {
  try await withCheckedThrowingContinuation { [unowned mobileClient] continuation in
    mobileClient.confirmSignIn(challengeResponse: code) { result, error in
      if let error {
        continuation.resume(throwing: error)
        return
      }
      
      guard let result, let loginResult = LoginResult.from(result: result) else {
        continuation.resume(throwing: .emptyResponse)
        return
      }
      
      continuation.resume(returning: loginResult)
    }
  }
}
```
And that works just fine. Maybe this is considered an "off-label" use of TOTP here but it works. 

Is there another way we should be going about this? And if AWSMobileClient doesn't support TOTP should we be looking into upgrading to the Amplify libraries if we require this functionality?

**Comment by harsh62** [comment]
So from your code snippet I can see that you are using `AWSMobileClient`. Just to be clear `AWSMobileClient` and `Amplify` does not support the following challenges coming from Cognito
* `SELECT_MFA_TYPE`
* `MFA_SETUP`
* `SOFTWARE_TOKEN_MFA` - Unofficially supported because of an unintended code path.

You can use `AWSCognitoIdentityProvider` directly (similar to how you are using `setUserMFAPreference`) to implement and handle all challenges coming from Cognito.

Note: Amplify V2 is currently working on prioritizing the implementation of TOTP challanges coming from `Cognito`.

**Comment by harsh62** [comment]
@tylerjames Yes in order to achieve your use case, you would need to use `AWSCognitoIdentityProvider `. There is no official documentation around this. But I can suggest you to go through the [`AWSMobileClient` implementation](https://github.com/aws-amplify/aws-sdk-ios/tree/main/AWSAuthSDK/Sources/AWSMobileClient), which can give you a pretty good idea on how `AWSCognitoIdentityProvider` works.

Similar to how you are using `setUserMFAPreference`, you can use `getSession` 

```
 AWSCognitoIdentityProvider
        .default()
        .getSession(). // pass in the username and password
```

**Comment by tylerjames** [comment]
Hmm, that's interesting.

If we use this method to log in would AWSMobileClient still "know" that we have successfully logged in?

Asking because we use the mobile client in a couple spots such as a "Change Password" feature inside the My Account screen. We're also using Kinesis which requires the AWSCredentials returned from `mobileClient.getAWSCredentials` to generate a signed URL.

**Comment by harsh62** [comment]
So `AWSCognitoIdentityProvider` and `AWSMobileCient` use different ways of securing the credentials in the Keychain. Unfortunately, we cannnot guarantee that both of them would work without any issue when using them independently in the same project.

**Comment by tylerjames** [comment]
Okay, I'll give it a shot and see how it goes. There might be ways around the mobile client use in the other parts of the app. I just don't want to have to manage all the tokens manually or something like that
